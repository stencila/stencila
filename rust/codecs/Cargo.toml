[package]
name = "codecs"
version = "0.0.0"
edition = "2021"

[features]
proptest-min = ["codec/proptest-min"]
proptest-low = ["codec/proptest-low"]
proptest-high = ["codec/proptest-high"]
proptest-max = ["codec/proptest-max"]

[dependencies]
codec = { path = "../codec" }
codec-cbor = { path = "../codec-cbor" }
codec-debug = { path = "../codec-debug" }
codec-html = { path = "../codec-html" }
codec-jats = { path = "../codec-jats" }
codec-json = { path = "../codec-json" }
codec-json5 = { path = "../codec-json5" }
codec-jsonld = { path = "../codec-jsonld" }
codec-markdown = { path = "../codec-markdown" }
codec-text = { path = "../codec-text" }
codec-yaml = { path = "../codec-yaml" }
node-strip = { path = "../node-strip" }

[dev-dependencies]
common-dev = { path = "../common-dev" }
divan = "0.1.5"
json_value_merge = "2.0.0"

[[bench]]
name = "from_str"
harness = false

[[bench]]
name = "from_bytes"
harness = false

[[bench]]
name = "to_string"
harness = false

[[bench]]
name = "to_bytes"
harness = false
