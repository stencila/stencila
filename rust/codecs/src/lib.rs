use std::path::{Path, PathBuf};

use codec::{
    common::{
        eyre::{bail, eyre, Result},
        reqwest::Client,
        tracing,
    },
    schema::Node,
};
pub use codec::{
    format::Format, Codec, CodecDirection, CodecSupport, DecodeInfo, DecodeOptions, EncodeInfo,
    EncodeOptions, Losses, LossesResponse, Mapping, MappingEntry, Message, MessageLevel, Messages,
    PoshMap, Position16, Position8, Positions, Range16, Range8,
};
use node_strip::{StripNode, StripTargets};

pub mod cli;

/// Get a list of all codecs
pub fn list() -> Vec<Box<dyn Codec>> {
    let codecs = vec![
        Box::new(codec_cbor::CborCodec) as Box<dyn Codec>,
        Box::new(codec_debug::DebugCodec),
        Box::new(codec_docx::DocxCodec),
        Box::new(codec_dom::DomCodec),
        Box::new(codec_directory::DirectoryCodec),
        Box::new(codec_html::HtmlCodec),
        Box::new(codec_jats::JatsCodec),
        Box::new(codec_json::JsonCodec),
        Box::new(codec_json5::Json5Codec),
        Box::new(codec_jsonld::JsonLdCodec),
        Box::new(codec_latex::LatexCodec),
        Box::new(codec_markdown::MarkdownCodec),
        Box::new(codec_odt::OdtCodec),
        Box::new(codec_pandoc::PandocCodec),
        Box::new(codec_pdf::PdfCodec),
        Box::<codec_swb::SwbCodec>::default(),
        Box::new(codec_text::TextCodec),
        Box::new(codec_yaml::YamlCodec),
    ];

    // TODO: make plugins a dependency and append codecs to list
    //let provided_by_plugins = &mut plugins::codecs::list();
    //codecs.append(provided_by_plugins);

    codecs
}

/// Resolve whether an optional string is a codec
pub fn codec_maybe(name: &str) -> Option<String> {
    list()
        .iter()
        .any(|codec| codec.name() == name)
        .then(|| name.to_string())
}

/// Get the codec for a given format
pub fn get(
    name: Option<&String>,
    format: Option<&Format>,
    direction: Option<CodecDirection>,
) -> Result<Box<dyn Codec>> {
    if let Some(name) = name {
        list()
            .into_iter()
            .find_map(|codec| (codec.name() == name).then_some(codec))
            .ok_or_else(|| eyre!("Unable to find a codec with name `{name}`"))
    } else if let Some(format) = format {
        list()
            .into_iter()
            .find_map(|codec| {
                match direction {
                    Some(CodecDirection::Decode) => {
                        codec.supports_from_format(format).is_supported()
                    }
                    Some(CodecDirection::Encode) => codec.supports_to_format(format).is_supported(),
                    None => {
                        codec.supports_from_format(format).is_supported()
                            || codec.supports_to_format(format).is_supported()
                    }
                }
                .then_some(codec)
            })
            .ok_or_else(|| eyre!("Unable to find a codec supporting format `{format}`"))
    } else {
        bail!("One of `name` or `format` must be supplied")
    }
}

/// Decode a Stencila Schema node from a string
#[tracing::instrument]
pub async fn from_str(str: &str, options: Option<DecodeOptions>) -> Result<Node> {
    let (node, DecodeInfo { losses, .. }) = from_str_with_info(str, options.clone()).await?;
    if !losses.is_empty() {
        let options = options.unwrap_or_default();
        let format = options
            .format
            .map(|format| format!("{format} ", format = format.name()))
            .unwrap_or_default();
        losses.respond(
            format!("While decoding from {format}string"),
            options.losses,
        )?;
    }

    Ok(node)
}

/// Decode a Stencila Schema node from a string with decoding losses
#[tracing::instrument]
pub async fn from_str_with_info(
    str: &str,
    options: Option<DecodeOptions>,
) -> Result<(Node, DecodeInfo)> {
    let codec = options.as_ref().and_then(|options| options.codec.as_ref());

    let format = options
        .as_ref()
        .and_then(|options| options.format.clone())
        .unwrap_or(Format::Json);

    let codec = get(codec, Some(&format), Some(CodecDirection::Decode))?;

    codec
        .from_str(
            str,
            Some(DecodeOptions {
                format: Some(format),
                ..options.unwrap_or_default()
            }),
        )
        .await
}

/// Decode a Stencila Schema node from a file system path
#[tracing::instrument]
pub async fn from_path(path: &Path, options: Option<DecodeOptions>) -> Result<Node> {
    let (node, DecodeInfo { losses, .. }) = from_path_with_info(path, options.clone()).await?;
    if !losses.is_empty() {
        let options = options.unwrap_or_default();
        losses.respond(
            format!("While decoding from path `{path}`", path = path.display()),
            options.losses,
        )?;
    }

    Ok(node)
}

/// Decode a Stencila Schema node from a URL (http://, https://, or file://)
#[tracing::instrument]
pub async fn from_url(url: &str, options: Option<DecodeOptions>) -> Result<Node> {
    if url.starts_with("https://") || url.starts_with("http://") {
        // TODO: If a format or media type is specified in options than
        // use that, otherwise use the `Content-Type` header, otherwise
        // (or maybe if plain text / octet stream) then use path.
        // This is just a temporary hack
        let options = Some(DecodeOptions {
            format: Some(Format::Markdown),
            ..options.unwrap_or_default()
        });

        // TODO: Enable HTTP caching to avoid unnecessary requests
        let response = Client::new().get(url).send().await?;
        if let Err(error) = response.error_for_status_ref() {
            let message = response.text().await?;
            bail!("{error}: {message}")
        }

        let text = response.text().await?;
        from_str(&text, options).await
    } else if let Some(path) = url.strip_prefix("file://") {
        from_path(&PathBuf::from(path), options).await
    } else {
        bail!("unknown URL protocol: {url}")
    }
}

/// Decode a Stencila Schema node from a file system path with decoding losses
#[tracing::instrument]
pub async fn from_path_with_info(
    path: &Path,
    options: Option<DecodeOptions>,
) -> Result<(Node, DecodeInfo)> {
    if !path.exists() {
        bail!("Path does not exist: {}", path.display());
    }

    let codec = options.as_ref().and_then(|options| options.codec.as_ref());

    let format = match options.as_ref().and_then(|options| options.format.clone()) {
        Some(format) => format,
        None => Format::from_path(path),
    };

    let codec = get(codec, Some(&format), Some(CodecDirection::Decode))?;

    codec
        .from_path(
            path,
            Some(DecodeOptions {
                format: Some(format),
                ..options.unwrap_or_default()
            }),
        )
        .await
}

/// Decode a Stencila Schema node from `stdin`
#[tracing::instrument]
pub async fn from_stdin(options: Option<DecodeOptions>) -> Result<Node> {
    use std::io::{self, BufRead};

    let stdin = io::stdin();
    let mut content = String::new();
    for line in stdin.lock().lines() {
        content += &line?;
        content.push('\n'); // Need to add the newline back on (e.g for Markdown)
    }

    from_str(&content, options).await
}

/// Encode a Stencila Schema node to a string
#[tracing::instrument(skip(node))]
pub async fn to_string(node: &Node, options: Option<EncodeOptions>) -> Result<String> {
    let (content, EncodeInfo { losses, .. }) = to_string_with_info(node, options.clone()).await?;
    if !losses.is_empty() {
        let options = options.unwrap_or_default();
        let format = options
            .format
            .map(|format| format!("{format} ", format = format.name()))
            .unwrap_or_default();
        losses.respond(format!("While encoding to {format}string"), options.losses)?;
    }

    Ok(content)
}

/// Encode a Stencila Schema node to a string with encoding losses
#[tracing::instrument(skip(node))]
pub async fn to_string_with_info(
    node: &Node,
    options: Option<EncodeOptions>,
) -> Result<(String, EncodeInfo)> {
    let codec = options.as_ref().and_then(|options| options.codec.as_ref());

    let format = options
        .as_ref()
        .and_then(|options| options.format.clone())
        .unwrap_or(Format::Json);

    let codec = get(codec, Some(&format), Some(CodecDirection::Encode))?;

    let options = Some(EncodeOptions {
        format: Some(format),
        ..options.unwrap_or_default()
    });

    if let Some(EncodeOptions {
        strip_scopes,
        strip_types,
        strip_props,
        ..
    }) = options.clone()
    {
        if !(strip_scopes.is_empty() && strip_types.is_empty() && strip_props.is_empty()) {
            let mut node = node.clone();
            node.strip(&StripTargets::new(strip_scopes, strip_types, strip_props));
            return codec.to_string(&node, options).await;
        }
    }

    codec.to_string(node, options).await
}

/// Encode a Stencila Schema node to a file system path
#[tracing::instrument(skip(node))]
pub async fn to_path(node: &Node, path: &Path, options: Option<EncodeOptions>) -> Result<()> {
    let EncodeInfo { losses, .. } = to_path_with_info(node, path, options.clone()).await?;
    if !losses.is_empty() {
        losses.respond(
            format!("While encoding to `{path}`", path = path.display()),
            options.unwrap_or_default().losses,
        )?;
    }

    Ok(())
}

/// Encode a Stencila Schema node to a file system path with losses
#[tracing::instrument(skip(node))]
pub async fn to_path_with_info(
    node: &Node,
    path: &Path,
    options: Option<EncodeOptions>,
) -> Result<EncodeInfo> {
    let codec = options.as_ref().and_then(|options| options.codec.as_ref());

    let format = match options.as_ref().and_then(|options| options.format.clone()) {
        Some(format) => format,
        None => Format::from_path(path),
    };

    let codec = get(codec, Some(&format), Some(CodecDirection::Encode))?;

    let options = Some(EncodeOptions {
        format: Some(format),
        ..options.unwrap_or_default()
    });

    if let Some(EncodeOptions {
        strip_scopes,
        strip_types,
        strip_props,
        ..
    }) = options.clone()
    {
        if !(strip_scopes.is_empty() && strip_types.is_empty() && strip_props.is_empty()) {
            let mut node = node.clone();
            node.strip(&StripTargets::new(strip_scopes, strip_types, strip_props));
            return codec.to_path(&node, path, options).await;
        }
    }

    codec.to_path(node, path, options).await
}

/// Convert a document from one format to another
#[tracing::instrument]
pub async fn convert(
    input: Option<&Path>,
    output: Option<&Path>,
    decode_options: Option<DecodeOptions>,
    encode_options: Option<EncodeOptions>,
) -> Result<String> {
    let node = match input {
        Some(input) => from_path(input, decode_options).await?,
        None => from_stdin(decode_options).await?,
    };

    match output {
        Some(output) => {
            to_path(&node, output, encode_options).await?;
            Ok(String::new())
        }
        None => to_string(&node, encode_options).await,
    }
}
