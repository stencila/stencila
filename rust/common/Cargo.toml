[package]
name = "common"
version = "0.0.0"
edition = "2021"

[dependencies]
Inflector = "0.11.4"
async-recursion = "1.0.5"
async-trait = "0.1.77"
base64 = "0.21.7"
bs58 = "0.5.0"
chrono = { version = "0.4.34", features = ["serde"] }
clap = { version = "4.5.1", features = ["derive", "color"] }
derivative = "2.2.0"
derive_more = "0.99.17"
dirs = "5.0.1"
eyre = "0.6.12"
futures = "0.3.30"
glob = "0.3.1"
indexmap = { version = "2.2.3", features = ["serde"] }
itertools = "0.12.1"
maplit = "1.0.2"
once_cell = "1.19.0"
proc-macro2 = "1.0.78"
quote = "1.0.35"
regex = "1.10.3"
reqwest = { version = "0.11.24", features = ["json"] }
serde = { version = "1.0.197", features = ["derive", "rc"] }
serde_json = { version = "1.0.114", features = ["preserve_order"] }
serde_with = "3.6.1"
serde_yaml = "0.9.32"
similar = { version = "2.4.0", features = ["unicode"] }
slug = "0.1.5"
smart-default = "0.7.1"
smol_str = "0.2.1"
strum = { version = "0.26.1", features = ["derive"] }
syn = "2.0.51"
tar = "0.4.40"
tempfile = "3.10.1"
tokio = { version = "1.36.0", features = ["full", "tracing"] }
toml = "0.8.10"
tracing = { version = "0.1.40", features = ["log"] }
type-safe-id = "0.2.1"
uuid = "1.7.0"
xz2 = "0.1.7"
zip = "0.6.6"
