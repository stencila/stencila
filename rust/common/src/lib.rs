//! # Common dependencies
//!
//! This internal crate simply re-exports dependencies that are commonly used across other internal
//! crates. Rust has a small `std` library (["and that's OK"](https://blog.nindalf.com/posts/rust-stdlib/)).
//! This crate acts as our internal version of a standard library, like [`stdx`](https://github.com/brson/stdx)
//! and others.
//!
//! The primary benefit of this crate is that there is only one place that version numbers for
//! commonly used dependencies need to be updated. Some of these crates are in line to become
//! part of the `std` library (e.g. `once_cell`).

pub use async_recursion;
pub use async_trait;
pub use base64;
pub use bs58;
pub use chrono;
pub use clap;
pub use derivative;
pub use derive_more;
pub use dirs;
pub use eyre;
pub use futures;
pub use glob;
pub use indexmap;
pub use inflector;
pub use itertools;
pub use maplit;
pub use once_cell;
pub use proc_macro2;
pub use quote;
pub use rand;
pub use regex;
pub use reqwest;
pub use serde;
pub use serde_json;
pub use serde_with;
pub use serde_yaml;
pub use similar;
pub use slug;
pub use smart_default;
pub use smol_str;
pub use strum;
pub use syn;
pub use tar;
pub use tempfile;
pub use tokio;
pub use toml;
pub use tracing;
pub use type_safe_id;
pub use uuid;
pub use which;
pub use xz2;
pub use zip;
