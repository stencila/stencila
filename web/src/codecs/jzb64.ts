import { Entity } from '@stencila/types'

/**
 * Decode `jzb64` (JSON Zipped and Base64 encoded) into a Stencila node
 *
 * This is a browser based implementation of `from_jzb64` in
 * rust/node-url/src/lib.rs.
 */
export async function decode(jzb64: string): Promise<Entity> {
  let standardBase64 = jzb64.replace(/-/g, '+').replace(/_/g, '/')
  while (standardBase64.length % 4) {
    standardBase64 += '='
  }

  const compressedData = atob(standardBase64)

  const uint8Array = new Uint8Array(compressedData.length)
  for (let i = 0; i < compressedData.length; i++) {
    uint8Array[i] = compressedData.charCodeAt(i)
  }

  const decompressedStream = new DecompressionStream('deflate')
  const writer = decompressedStream.writable.getWriter()
  const reader = decompressedStream.readable.getReader()

  writer.write(uint8Array)
  writer.close()

  const { value } = await reader.read()
  const jsonString = new TextDecoder().decode(value)
  return JSON.parse(jsonString)
}
