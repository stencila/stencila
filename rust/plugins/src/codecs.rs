use std::sync::Arc;

use codec::{Codec, EncodeInfo, EncodeOptions, schema::Node, status::Status};
use common::{
    async_trait::async_trait,
    eyre::{Result, bail},
    serde::{Deserialize, Serialize},
    tokio::sync::Mutex,
};

use crate::{Plugin, PluginInstance, plugins};

/// A codec provided by a plugin
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(crate = "common::serde")]
pub struct PluginCodec {
    /// The name of the codec
    name: String,

    /// The plugin that provides this codec
    ///
    /// Used to be able to create a plugin instance, which in
    /// turn is used to create a codec instance.
    #[serde(skip)]
    plugin: Option<Plugin>,

    /// The plugin instance for this codec. Used to avoid starting
    /// a new instance for each call to the codec.
    ///
    /// This needs to be a `Arc<Mutex>` because the `to_string`, `from_string` etc methods are async
    /// but not `&mut self`. As such, this is needed for "interior mutability" across
    /// calls to those methods.
    #[serde(skip)]
    plugin_instance: Arc<Mutex<Option<PluginInstance>>>,
}

impl PluginCodec {
    /// Bind a plugin to this codec so that it can be started (by starting the plugin first)
    pub fn bind(&mut self, plugin: &Plugin) {
        self.plugin = Some(plugin.clone());
    }
}

#[async_trait]
impl Codec for PluginCodec {
    fn name(&self) -> &str {
        &self.name
    }

    fn status(&self) -> Status {
        Status::Alpha
    }

    async fn to_string(
        &self,
        node: &Node,
        _options: Option<EncodeOptions>,
    ) -> Result<(String, EncodeInfo)> {
        // Create the plugin instance if necessary
        let mut guard = self.plugin_instance.lock().await;
        let instance = match &mut *guard {
            Some(instance) => instance,
            None => {
                let Some(plugin) = self.plugin.as_ref() else {
                    bail!("Not bound yet")
                };

                let inst = plugin.start(None).await?;
                *guard = Some(inst);
                guard.as_mut().expect("should have been set above")
            }
        };

        // Call the plugin's to_string method
        #[derive(Serialize)]
        #[serde(crate = "common::serde")]
        struct Params<'node> {
            codec: String,
            node: &'node Node,
        }
        let result: String = instance
            .call(
                "codec_to_string",
                Params {
                    codec: self.name.clone(),
                    node,
                },
            )
            .await?;

        Ok((result, EncodeInfo::none()))
    }
}

/// List all the codecs provided by plugins
pub async fn list() -> Result<Vec<Box<dyn Codec>>> {
    Ok(plugins()
        .await
        .into_iter()
        .flat_map(|plugin| plugin.codecs())
        .collect())
}
