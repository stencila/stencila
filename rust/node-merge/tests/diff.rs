use codecs::{DecodeOptions, Format};
use common::{eyre::Result, tokio};
use common_dev::insta::assert_debug_snapshot;

use node_merge::{DiffResult, MergeNode};

/// Snapshot tests of the `MergeNode::diff` method
#[tokio::test]
async fn diff() -> Result<()> {
    async fn diff(old: &str, new: &str) -> Result<DiffResult> {
        let options = Some(DecodeOptions {
            format: Some(Format::Markdown),
            ..Default::default()
        });
        let old = codecs::from_str(old, options.clone()).await?;
        let new = codecs::from_str(new, options).await?;

        Ok(old.diff(new))
    }

    macro_rules! diff {
        ($old:literal, $new:literal) => {
            diff($old, $new).await?
        };
    }

    assert_debug_snapshot!("same", diff!(r#"One"#, r#"One"#));

    assert_debug_snapshot!(
        "change-one-prop",
        diff!(
            r#"
```js
Code
```
"#,
            r#"
```py
Code
```
"#
        )
    );

    assert_debug_snapshot!(
        "change-all-props",
        diff!(
            r#"
```js
Code
```
"#,
            r#"
```py
Diff code
```
"#
        )
    );

    assert_debug_snapshot!(
        "change-node-type",
        diff!(
            r#"
```js
Code
```
"#,
            r#"
```py exec
Code
```
"#
        )
    );

    assert_debug_snapshot!(
        "missing-backticks",
        diff!(
            r#"
```js
Code
```
"#,
            r#"
```js
Code

more
"#
        )
    );


    Ok(())
}
