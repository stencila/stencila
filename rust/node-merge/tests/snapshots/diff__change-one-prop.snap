---
source: rust/node-merge/tests/diff.rs
expression: "diff!(r#\"\n```js\nCode\n```\n\"#, r#\"\n```py\nCode\n```\n\"#)"
---
art_11111112.cdb_11111111  Article:content.0.CodeBlock:code                 CodeBlock:code                 "Code"
art_11111112.cdb_11111111  Article:content.0.CodeBlock:programmingLanguage  CodeBlock:programmingLanguage  "js"

art_11111112.cdb_11111111  Article:content.0.CodeBlock:code                 CodeBlock:code                 "Code"
art_11111112.cdb_11111111  Article:content.0.CodeBlock:programmingLanguage  CodeBlock:programmingLanguage  "py"

DiffOp       Old range    New range
Equal        0..1         0..1
Delete       1..2         2..2
Insert       1..1         1..2

Replace((MathBlock:code, "foo"))
